### PR TITLE
feat(input): component api for input masking

### PR DESCRIPTION
**Pull request by sean-perkins**
Issue number: Internal

---------

 

 

## What is the current behavior?

N/A

## What is the new behavior?

- Adds `mask`, `maskVisibility` and `maskPlaceholder` properties to `ion-input` for input masking

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by sean-perkins** [comment]
This is one of many pull requests for adding input masking to `ion-input`. As functionality becomes available for user testing, I will call that out in the pull request. 

I will do my best to break down the work effort behind this initiative. There are going to be some complex chunks of work along the journey 🙂 

All individual PRs should merge back into `FW-2573` before the feature is eventually merged into a 7.1 feature branch.

File: core/src/components/input/test/input.spec.ts
Comment (amandaejohnston) [inline core/src/components/input/test/input.spec.ts:17]: Are these tests necessary? We're just checking if Stencil passed the correct default value here. I can't imagine the defaults changing by accident/without being caught in review.

Comment (sean-perkins) [reply core/src/components/input/test/input.spec.ts:17]: I don't know if any tests are necessary, but it helps avoid the potential for human error in both development and review and allows projects to iterate quicker with extra validation that existing behavior is maintained.

File: core/src/components/input/input.tsx
Comment (liamdebeasi) [inline core/src/components/input/input.tsx:280]:
```suggestion
   * constant characters. Each index of the mask array defines the mask for
```

File: core/src/components/input/input.tsx
Comment (liamdebeasi) [inline core/src/components/input/input.tsx:294]:
```suggestion
   * Character or string to cover unfilled parts of the mask. The default character is `_`.
```

File: core/src/components/input/input.tsx
Comment (liamdebeasi) [inline core/src/components/input/input.tsx:277]: What's the difference between the "string" and "character" terminology?

File: core/src/components/input/input.tsx
Comment (liamdebeasi) [inline core/src/components/input/input.tsx:283]: We should also document how certain characters define the allowed inputs

File: core/src/components/input/input.tsx
Comment (sean-perkins) [reply core/src/components/input/input.tsx:277]: I believe this is answered based on our Slack thread + design doc adjustment proposal, but let me know if that isn't the case. 

Once I have approval on the design doc adjustment, I'll update the type and description for this.

File: core/src/components/input/input.tsx
Comment (liamdebeasi) [inline core/src/components/input/input.tsx:285]: When I first read this I understood it to mean each entry in the array is a fixed character. This would contradict the above line which states the array can accept a regex. However, I now understand that it means string types.

Perhaps we can just say "The fixed characters in the mask cannot be erased or replaced by the user"

File: core/src/components/input/test/input.spec.ts
Comment (liamdebeasi) [reply core/src/components/input/test/input.spec.ts:17]: I don't mind keeping them, but I will note that any change to the default value of props will cause `api.txt` to be updated so we should be able to catch accidental changes to the default value during review.

**Review by liamdebeasi — APPROVED** [review]

**Comment by sean-perkins** [comment]
Sweet - in the case of the `api.txt` I'll remove the test for now and can re-evaluate as the feature is developed.

**Pull request merged**
